### PR TITLE
[WIP] Replace all audio backends with Cubeb

### PR DESCRIPTION
**Pull request by ligfx**
This is a big change, hear me out.

This PR removes all existing Dolphin audio backends, as well as the the PortAudio shim for microphones and the option to switch between backends, and replaces it all with a vendored copy of Cubeb. A large chunk of Dolphin code is removed (about 65,000 lines) and all audio features should Just Work™ everywhere.

Cubeb is the cross-platform real-time audio backend [developed for and used by Firefox](https://blog.mozilla.org/webrtc/firefoxs-audio-backend/). It gets exposure across the entire Firefox user base, works on every major platform, supports both input and output, and is available as [an independent library](https://github.com/kinetiknz/cubeb).

**Pros:**
- Volume control and DPL2 decoder supported on all platforms.
- No more maintaining six different audio output backends.
- No more bundling/linking to PortAudio.
- Defaults to using WASAPI on Windows, which is callback-based, lower-level than XAudio2 (no audio effects graph), and available since Vista.
- Removes a configuration option.
- Loads PulseAudio (and JACK, if you're into that) at runtime on Linux. If the shared library isn't found, it seamlessly falls back to ALSA.
- Also supports JACK and OpenBSD's sndio.
- Larger pool of people working on improvements and bug fixes for Cubeb, instead of relatively siloed Dolphin audio development.

**Cons:**
- No guarantee of a stable API/ABI, since Cubeb is built in-tree by Firefox and they can change it whenever they want.
- No more hardware OpenAL support (shouldn't be a big issue as OpenAL is being used as a dump pipe anyways).
- Cubeb internal code is more complex than Dolphin's current audio code, due to additional features they support (multichannel on every backend, full-duplex streams, software resampling on WASAPI, etc.).
- The ALSA backend is not a Mozilla priority (they stopped shipping it in Firefox), so may not receive updates or as many fixes.

**Currently tested:**
- [ ] Windows 7
- [ ] Windows 8
- [x] macOS: by @ligfx.
  - Ran Super Mario Sunshine for hours with no problems
  - Singing in Karaoke Revolution Party seemed the same as with PortAudio
  - Automatically switches output device with rest of system
- [ ] Linux ALSA
- [ ] Linux PulseAudio
- [ ] Android

### CODE REVIEW

**Comment by ligfx** [comment]
@MerryMage @degasus @Orphis @JMC47 etc

**Comment by Helios747** [comment]
That's... A heck of a proposal. I dunno how I feel about this. Especially considering the timestretch audio work going on which this will kill.


Huh. I dunno. I'm all for cutting down on the absurd number of backends we have.

**Comment by ligfx** [comment]
> Especially considering the timestretch audio work going on which this will kill.

The time-stretching logic has been moved into the Mixer class, so now automatically works with all backends, including this one!

**Comment by MayImilae** [comment]
I'm not a coder, but from my experience as a tester and from working with users, I'm against having a single API for output like this.

**Comment by ligfx** [comment]
@Maylmilae what specifically about having a single API are you against? I think it would overall be beneficial to users because it would bring feature-parity to all platforms and remove a trivial option choice.

**Comment by lex3a** [comment]
I don't think this will be merged anytime soon. However from a user perspective it looks great. I hate a lot of backends. But we will get a lot of regressions after this,  aren't we?

File: Source/Core/AudioCommon/CubebStream.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+{
+  auto self = static_cast<CubebStream*>(user_data);
+
+  if (self->m_stereo)
+  {
+    self->m_mixer->Mix(static_cast<short*>(outputbuffer), nframes);
+  }
+  else
+  {
+    if (static_cast<size_t>(nframes) * 2 > self->m_conversion_buffers_size)
```
Comment (MerryMage) [inline Source/Core/AudioCommon/CubebStream.cpp]: This could probably be simplified by using `std::vector` instead of `std::unique_ptr<[]>` and just calling `std::vector::reserve` unconditionally.

I don't have very strong opinions on this though.

**Comment by degasus** [comment]
I'm all for killing backends :D

**Comment by MerryMage** [comment]
From a code perspective this looks straightforward, since a dumb-pipe is all that we require.

Some questions:

* Should a log callback be registered? (Is this necessary?)
* It would be nice to emit samples at the prefered sample rate so the backend doesn't have to do interpolation again.

**Comment by MayImilae** [comment]
I'm still hesitant to have a single output, but, let's talk realistically about what we need to do if we're going to move entirely to Cubeb.

If we're going to remove all other audio backends all at once and replace them with Cubeb, if we're going to do this responsibly, we need to do this:

1. All operating systems need to be tested for compliance. Linux, Android, macOS, FreeBSD, Windows 7, 8.1, and 10.
2. Everything that can use sound needs to be tested. Gameplay, DTK audio, audio dumping, gameplay recording, streaming, TAS (just in case), etc etc. And this needs to be tested in every operating system, too.
3. Latency needs to be analyzed on all platforms.
4. Performance needs to be checked, just in case there is anything stupid on any platform (users will ALWAYS find performance regressions!)
5. Regression fixes after the fact. There are bound to be some.

Basically, this is a lot to test! Doing this all in one go is going to take a lot out of everyone!

How about a progressive rollout solution? Adding Cubeb, announcing our intentions on the progress report, and removing backends bit by bit? Or even announcing it in the progress report and removing the other backends the month after? Give users some time to try it out, and testers a chance to put it through its paces, before we get rid of all other options. It would make for an easier sell on the blog, there would be a chance for it to be out in the wild before the other options are lost, and we'd have a proper chance to test it.

**Comment by ligfx** [comment]
Repeating what I said on IRC for posterity:

@MayImilae I'm not completely against a staggered rollout, but the items in your point 2 are unaffected by this change. All of that goes through the Mixer. This PR only replaces the dump pipe at the very end, which is separated from the actual logic by the SoundStream interface.

**Comment by MerryMage** [comment]
Re @MayImilae:

Personally I would support a staggered rollout. It would be a less of a sudden change and gives the users an opportunity to voice any concerns about the change before it happens, as well as have a broader testing base before we commit to removing the other backends.

Even if there are no major breakages that we can foresee (the audio backends *are* isolated quite well), it would be nice to think about the users' perspective as well as unforeseen complications.

**Comment by degasus** [comment]
I agree to 1. and 3., but 2. and 4. are wrong. The audio backend has no impact on emulation, neither in terms of performance nor in terms of correctness. So if the backend manages to stream out well anything without stotters, this is fine.

**Comment by MayImilae** [comment]
@degasus It still needs to be done though for at least one reason: the blog, and selling it to users. Users are going to worry about performance and how well it works with X feature, so if we can say we tested _everything_ and can show results, it will really really help with user acceptance of this.

File: Externals/cubeb/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+  configure_file(${CMAKE_CURRENT_SOURCE_DIR}/docs/Doxyfile.in ${CMAKE_CURRENT_BINARY_DIR}/docs/Doxyfile @ONLY)
+  add_custom_target(doc ALL
+    ${DOXYGEN_EXECUTABLE} ${CMAKE_CURRENT_BINARY_DIR}/docs/Doxyfile
+    WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/docs
+    COMMENT "Generating API documentation with Doxygen" VERBATIM)
+endif()
+
+enable_testing()
+
+add_executable(test_sanity test/test_sanity.cpp)
```
Comment (Orphis) [inline Externals/cubeb/CMakeLists.txt]: There are lots of test executables that are going to be added to the solution. We should remove them or add a switch so they don't get added when cubeb is bundled in another project.

File: Source/Core/AudioCommon/CubebStream.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+#include "Core/ConfigManager.h"
+
+#include "cubeb/cubeb.h"
+
+constexpr u32 BUFFER_SAMPLES = 512;  // ~10 ms - needs to be at least 240 for surround
+
+long CubebStream::data_callback(cubeb_stream* stream, void* user_data, const void* /*inputbuffer*/,
+                                void* outputbuffer, long nframes)
+{
+  auto self = static_cast<CubebStream*>(user_data);
```
Comment (lioncash) [inline Source/Core/AudioCommon/CubebStream.cpp]: `auto*`

File: Source/Core/AudioCommon/CubebStream.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  return nframes;
+}
+
+void CubebStream::state_callback(cubeb_stream* stream, void* user_data, cubeb_state state)
+{
+}
+
+bool CubebStream::Start()
+{
+  if (cubeb_init(&m_ctx, "Dolphin", NULL) != CUBEB_OK)
```
Comment (lioncash) [inline Source/Core/AudioCommon/CubebStream.cpp]: `nullptr`

File: Source/Core/AudioCommon/CubebStream.cpp [hunk 5]
```diff
@@ -0,0 +1,112 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "AudioCommon/CubebStream.h"
+#include "AudioCommon/DPL2Decoder.h"
+#include "Common/Logging/Log.h"
```
Comment (lioncash) [inline Source/Core/AudioCommon/CubebStream.cpp]: `Common/CommonTypes.h` needs to be included.

File: Source/Core/AudioCommon/CubebStream.h [hunk 6]
```diff
@@ -0,0 +1,29 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
```
Comment (lioncash) [inline Source/Core/AudioCommon/CubebStream.h]: `<cstddef>` and `<memory>` needs to be included.

File: Source/Core/AudioCommon/CubebStream.h [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+private:
+  bool m_stereo;
+  cubeb* m_ctx;
+  cubeb_stream* m_stream;
+
+  size_t m_conversion_buffers_size{0};
+  std::unique_ptr<short[]> m_short_buffer;
+  std::unique_ptr<float[]> m_floatstereo_buffer;
+
+  static long data_callback(cubeb_stream* stream, void* user_data, const void* /*inputbuffer*/,
```
Comment (lioncash) [inline Source/Core/AudioCommon/CubebStream.h]: These should be named `DataCallback` and `StateCallback`

File: Source/Core/AudioCommon/CubebStream.h [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+                            void* outputbuffer, long nframes);
```
Comment (lioncash) [inline Source/Core/AudioCommon/CubebStream.h]: `output_buffer`, `num_frames`.

File: Source/Core/AudioCommon/CubebStream.h [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+class CubebStream final : public SoundStream
+{
+public:
+  bool Start() override;
+  void Stop() override;
+  void SetVolume(int) override;
+
+private:
+  bool m_stereo;
```
Comment (lioncash) [inline Source/Core/AudioCommon/CubebStream.h]: These should have initializers as well.

File: Source/Core/AudioCommon/CubebStream.cpp [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+#include "Common/Thread.h"
+#include "Core/ConfigManager.h"
+
+#include "cubeb/cubeb.h"
```
Comment (lioncash) [inline Source/Core/AudioCommon/CubebStream.cpp]: This should be using `<>` for inclusion.

File: Source/Core/AudioCommon/CubebStream.cpp [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+constexpr u32 BUFFER_SAMPLES = 512;  // ~10 ms - needs to be at least 240 for surround
+
+long CubebStream::data_callback(cubeb_stream* stream, void* user_data, const void* /*inputbuffer*/,
+                                void* outputbuffer, long nframes)
```
Comment (lioncash) [inline Source/Core/AudioCommon/CubebStream.cpp]: `output_buffer`, `num_frames`

File: Source/Core/AudioCommon/CubebStream.cpp [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+
+#include "AudioCommon/CubebStream.h"
+#include "AudioCommon/DPL2Decoder.h"
+#include "Common/Logging/Log.h"
+#include "Common/Thread.h"
+#include "Core/ConfigManager.h"
+
+#include "cubeb/cubeb.h"
+
+constexpr u32 BUFFER_SAMPLES = 512;  // ~10 ms - needs to be at least 240 for surround
```
Comment (lioncash) [inline Source/Core/AudioCommon/CubebStream.cpp]: Please place comments above code where applicable.

File: Source/Core/AudioCommon/CubebStream.cpp [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+    params.format = CUBEB_SAMPLE_FLOAT32NE;
+    params.layout = CUBEB_LAYOUT_3F2_LFE;
+  }
+
+  u32 minimum_latency = 0;
+  if (cubeb_get_min_latency(m_ctx, params, &minimum_latency) != CUBEB_OK)
+    ERROR_LOG(AUDIO, "Error getting minimum latency");
+  INFO_LOG(AUDIO, "Minimum latency: %i frames", minimum_latency);
+
+  if (cubeb_stream_init(m_ctx, &m_stream, "Dolphin Audio Output", NULL, NULL, NULL, &params,
```
Comment (lioncash) [inline Source/Core/AudioCommon/CubebStream.cpp]: nullptr

File: Source/Core/AudioCommon/CubebStream.h [hunk 14]
```diff
@@ -0,0 +1,29 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include "AudioCommon/SoundStream.h"
+#include "cubeb/cubeb.h"
```
Comment (lioncash) [inline Source/Core/AudioCommon/CubebStream.h]: This should be using `<>` for inclusion.

File: Externals/cubeb/CMakeLists.txt (hunk 2 above)
Comment (ligfx) [inline Externals/cubeb/CMakeLists.txt]: I'll delete everything past line 149 for now.

File: Source/Core/AudioCommon/CubebStream.cpp (hunk 5 above)
Comment (ligfx) [inline Source/Core/AudioCommon/CubebStream.cpp]: Good eyes. Done.

File: Source/Core/AudioCommon/CubebStream.cpp (hunk 10 above)
Comment (ligfx) [inline Source/Core/AudioCommon/CubebStream.cpp]: Doh. Done.

File: Source/Core/AudioCommon/CubebStream.cpp (hunk 12 above)
Comment (ligfx) [inline Source/Core/AudioCommon/CubebStream.cpp]: Done.

File: Source/Core/AudioCommon/CubebStream.cpp (hunk 11 above)
Comment (ligfx) [inline Source/Core/AudioCommon/CubebStream.cpp]: Done.

File: Source/Core/AudioCommon/CubebStream.cpp (hunk 3 above)
Comment (ligfx) [inline Source/Core/AudioCommon/CubebStream.cpp]: Done.

File: Source/Core/AudioCommon/CubebStream.cpp (hunk 4 above)
Comment (ligfx) [inline Source/Core/AudioCommon/CubebStream.cpp]: Done.

File: Source/Core/AudioCommon/CubebStream.cpp (hunk 13 above)
Comment (ligfx) [inline Source/Core/AudioCommon/CubebStream.cpp]: Done.

File: Source/Core/AudioCommon/CubebStream.h (hunk 6 above)
Comment (ligfx) [inline Source/Core/AudioCommon/CubebStream.h]: Done.

File: Source/Core/AudioCommon/CubebStream.h (hunk 14 above)
Comment (ligfx) [inline Source/Core/AudioCommon/CubebStream.h]: Done.

File: Source/Core/AudioCommon/CubebStream.h (hunk 9 above)
Comment (ligfx) [inline Source/Core/AudioCommon/CubebStream.h]: Okay, I gave them all default initializers, does that work?

File: Source/Core/AudioCommon/CubebStream.h (hunk 7 above)
Comment (ligfx) [inline Source/Core/AudioCommon/CubebStream.h]: Done.

File: Source/Core/AudioCommon/CubebStream.h (hunk 8 above)
Comment (ligfx) [inline Source/Core/AudioCommon/CubebStream.h]: Done.

File: Source/Core/AudioCommon/CubebStream.cpp (hunk 1 above)
Comment (ligfx) [inline Source/Core/AudioCommon/CubebStream.cpp]: You're right, that is simpler. Changed.

**Comment by lioncash** [comment]
Not sure if it matters, but I fixed two minor bugs in cubeb today, so you may want to update to the most recent master or something if it does turn out to matter (one is a leak fix, and one fixes a bug on WASAPI).

**Comment by Fallcrest** [comment]
Will this still work with future plans for Dolphin expansion?

By that I mean, will it work when the iOS/tvOS builds and the Haiku build become real things (eventually), and will it work for when Dolphin starts doing VR (does it support positional audio, just in case that is a thing)?

**Comment by ligfx** [comment]
@lioncash Cool! I updated locally, I'll try to keep this PR up-to-date with any small fixes like that.

@Fallcrest I can't speak to any of those possibilities.

@merrymage I implemented handling Cubeb's logs, but it's a handful of icky code to integrate with LogManager. The logs are basically, "Stream started with $params, mixing to device format $param. ... Stream stopped." If you think that's useful, I can add it in to this PR. What do you think?

@Maylimilae @JMC47 *macOS testing:* I ran Super Mario Sunshine on macOS today for two hours. There were no crashes and the audio did not exhibit any latency or quality changes (checked at the beginning and end of the two hours).

**Comment by mbc07** [comment]
`Will this still work with future plans for Dolphin expansion?`

Wouldn't this be a matter of just adding another audio backend if Cubeb can't handle those new cases when Dolphin becomes ready for those expansions?

**Comment by Fallcrest** [comment]
@mbc07 Fair point, but removing things now when they might be needed later would slow development of those potential new features down the line. I'm all for having only one audio backend in the long run, but it is important to make sure it will serve all of Dolphin's needs in the near future if the idea is to remove everything except that.

@waddlesplash would have to see whether or not this will work for the work on the Haiku build. And I guess @grp on the iOS build. And if they don't work, perhaps issues could be fixed within Cubeb so that only one backend is necessary.

**Comment by Helios747** [comment]
A platform with 2 users and a platform not even merged for support is not what I would call a blocker on this, Fallcrest.

**Comment by waddlesplash** [comment]
We don't have a cubeb port yet, but from a quick glance at its internal architecture, a port should take a weekend or so for someone familiar with Haiku's audio system. Not a huge deal (and yeah, not a blocker by any means.)

**Comment by Fallcrest** [comment]
I never said it would be a blocker on this, I just thought it would be a good idea to hear from the relevant people with more unique needs in order to know as much as we can beforehand.

@waddlesplash, good to know a port wouldn't require much effort.

**Comment by MerryMage** [comment]
@ligfx For logs I was mainly worried about error logs. If cubeb doesn't produce such things, I guess don't bother?

**Comment by ligfx** [comment]
@MerryMage So I looked into it further, Cubeb should produce error messages to that callback, if they happen. Two issues I'm running into with implementation:
- Cubeb has no concept of log levels. The same callback is used for informational messages (like "Cubeb stream init successful.") and problem messages (like "underrun in cubeb backend." or "Could not create resampler."). I guess Dolphin could treat them all as errors, so they're always visible?
- The filename and line number are included in the message. That added on to the Dolphin filename and line number makes a _really_ long line. It's pretty easy to extract them and pass them manually, though that could be considered sort of hacky. They could also be passed straight through without Dolphin's standard formatting.

Do you or anyone else have any thoughts on those?

**Comment by ligfx** [comment]
I extracted the first part of this into #5313: Add CubebStream output backend and replace PortAudio for microphones

**Comment by Ofunniku** [comment]
My only real concern is if the "Enhancements" audio processing features in the Windows audio stack are not applying to the audio.  I rely on the Loudness Equalization, Bass Management features for sound output most of the time, and Noise Suppression, and Echo Cancellation for microphone input at all times.

Those still work as of 5.0-3939 with Cubeb output, but I have yet to test input, since my mic is busted.  I'm all for this change.

**Comment by shuffle2** [comment]
I debugged the assert I was getting from cubeb.  
Each time the EmuThread starts a game, dolphin calls `CubebStream::Start`. The first time cubeb starts up and it writes a log message,  this code will be hit:
```c++
void cubeb_async_log(char const * fmt, ...)
...
cubeb_async_logger::get().push(msg);
```
where `get` initializes and returns a singleton:
```c++
static cubeb_async_logger & get() {
    static cubeb_async_logger instance;
    return instance;
  }
```
This singleton starts a detached `std::thread` in `cubeb_async_logger::run` which is the log "consumer", and will stay around forever. The `cubeb_async_logger::msg_queue` is a `ring_buffer_base`. `ring_buffer_base` internally tracks the consumer and producer thread ids and expects them to be constant.

On successive game starts, an assert is hit because the "producer" thread id has changed: https://github.com/dolphin-emu/dolphin/blob/master/Externals/cubeb/src/cubeb_ringbuffer.h#L317

This makes perfect sense because cubeb's `wasapi_stream_start` creates a new thread to act as a producer each time `cubeb_stream_start` is called (and it's killed by `cubeb_stream_stop` in `cubeb_stream_start`).

Sooooo. This is a cubeb bug. Their log consumer is expecting thread ids to be constant, but cubeb itself is freely creating and killing producers (while the consumer is an ever-running singleton...).

**Comment by shuffle2** [comment]
maybe time for a rebase of this?

**Comment by ligfx** [comment]
@shuffle2 maybe. This PR doesn't seem relevant until cubeb has had a bit more testing (it was _just_ mentioned in the progress report), and I'm thinking of breaking off more interesting pieces of this anyways. How does "Drop OpenAL since Cubeb is cross-platform" sound?

**Comment by shuffle2** [comment]
That sounds fine to me. At least we can hear what complaints people have, then :)  
I was really just referring to the fact that it's hard to see what this PR is doing, now, since there are a lot of conflicts.

**Comment by LAGonauta** [comment]
I do not quite agree on the removal of OpenAL as it is an easy way to get HRTF on any platform (which will be worth to use when PR#5235 with FreeSurround is merged).

OpenAL can also have acceptable latency by setting "period_size=64" and "periods=2" on OpenAL Soft's configuration file (I did not search for the best values, just tried these and they worked), and by setting latency to "2" on Dolphin (which should be an additional 10 ms of delay on top of WASAPI). Of course one can also always use an X-Fi on Windows with hardware OpenAL where the latency goes from acceptable to great.

**Comment by ligfx** [comment]
@LAGonauta How do you envision HRTF being used? It looks to me like any positional audio data is lost deeper inside the DSP emulation, and not available to the audio output system.

**Comment by LAGonauta** [comment]
@ligfx Yes, all positional audio data is lost on DSP-HLE but it is not lost with both proprietary and free DSP ROMs when using DSP-LLE. FreeSurround can decode this data into 5.1/7.1 channels which then can be mixed back to stereo for headphones or speakers using HRTF. Works really well with OpenAL Soft and a X-Fi.

**Comment by ligfx** [comment]
Oh, you're talking about the 5.1-channel audio data. I was thinking about true positional 3d audio.

How does using HRTF with 5.1 audio sound? Is it like you're sitting in a living room with a surround sound setup?

I think most gaming headphones will do this automatically, and it also looks like the Windows 10 Creators Update can now do that as well. Have you tried their option?

**Comment by LAGonauta** [comment]
@ligfx Exactly, this is how it sounds like. FreeSurround can allow true-er positional audio if we allow it to decode into 18 channels and pass all channels to OpenAL, but I am not sure how it should sound like. Dolby Atoms-like? 
Not sure if most gaming headphones do, but some will do this automagically, yes. (Probably most only on Windows with custom drivers)
If you want I can help you setup OpenAL Soft so you can hear it yourself.

I tried the new auralizer (?) on W10 CU, Windows Spacial (forgot the name) has too much echo for me. Dolby Atmos sounds ok (but costs money and my evaluation time has already passed). Also, for some reason both worked only on Store apps when I tested, I did not have luck with MPC or any other sound source.
Besides, I do not know if Windows shows itself as a 7.1 sink when using it, I could test but right now my W10 machine is unavailable.
Most programs nowadays can only output surround sound if surround is set on Windows, but obviously it won't be if the auralizer (?) is enabled UNLESS Windows says the sink is 7.1 instead of headphones/stereo to the program.
Sadly I believe Cubeb is one of those (for now), even on Linux. At least I could not get it to output to a PulseAudio virtual sink (system-wide HRTF) on Linux (stereo sound card, 5.1 virtual sink with Kemar HRTF), while OpenAL not only was able to output to the virtual sink but I could also use it to mix using HRTF (app-wide HRTF).

**Comment by ligfx** [comment]
@LAGonauta I'm a little confused about something you said, how can FreeSurround decode 18 channels from a 5.1-channel stream?

**Comment by LAGonauta** [comment]
@ligfx Actually it decodes 16 channels (not 18, I was mistaken) from DPLII stereo encoded content. I read more and it is not Atmos-like (no height information). It decodes all channels in a circle around the listener so the transition between two speakers is smoother as they are close together, and should be more realistic. As HRTF allows and arbitrary number of speakers this should make the sound wrap around the listener with no noticeable transition between each sound source. Of course this probably depends on how Nintendo chose to encode the sound.

From a post by "pro_optimizer", FreeSurround's developer:

> The sound field right after the decoding stage can be pictured as a 2-dimensional square with the listener at its center. (...) The second group of controls is how the sound field is mapped onto the speakers. The first setting here is the channel setup; this allows you to select how many speakers you have and where they are placed (front left and right, center, rear left and right, side speakers, etc.). From the point of view of the decoder there at 16 possible positions along the edges of the square sound field (see [this](https://en.wikipedia.org/wiki/File:5_1_channels_(surround_sound)_label.svg) picture for an example) namely the 4 corners, the 4 midpoints (front center, back center, left side, and right side), and two positions on each edge half-way between the corners and the midpoint (e.g., front left center, front right center). Most people have only a subset of all these channels (e.g., 7.1) in some common locations, but you can have some fairly arbitrary setups (***). But note that the placement of the speakers in the room is actually not in a rectangle but instead follows the corresponding home theater rules. If you have a subwoofer you might want to do the bass management right here, but it is usually a better idea to let the sound card or amplifier do it if they offer the option (e.g. for active sub-satellite systems). (...)

**Comment by ligfx** [comment]
Gonna close this, removals in the future can be done one-by-one in individual PRs.

**Comment by shuffle2** [comment]
hm...well, it will be sad if cubeb started out as a way to remove backends, and instead just added one :-)

**Comment by shuffle2** [comment]
Seems they finally fixed one of the blocking bugs https://github.com/kinetiknz/cubeb/pull/350
